juriscraper/opinions/united_states/federal_bankruptcy/bap1.py

```diff
@@ -0,0 +1,124 @@
+
+    def _download(self, request_dict: Dict = {}) -> HtmlElement:
+        return super()._download(self.parameters)
+
+    def _process_html(self) -> None:
+        for row in self.html.xpath("//tr[td]"):
+            opinion_name = row.xpath("td[1]")[0].text_content().strip()
+            status = self.get_status_from_opinion_name(opinion_name)
+
+            lower_court = row.xpath("td[4]/span")[0].text_content().strip()
+            docket_number = row.xpath("td[2]")[0].text_content().strip()
+
+            case = {
+                "status": status,
+                "url": row.xpath("td[1]/a/@href")[0],  # opinion url
+                "docket": self.build_full_docket_number(
+                    docket_number, lower_court
+                ),
+                # Pub Date
+                "date": row.xpath("td[3]")[0].text_content().strip(),
+                # short title
+                "name": row.xpath("td[4]")[0].text.strip(),
+                # district
+                "lower_court": lower_court,
+            }
+
+            self.cases.append(case)
+
+    def set_date_filter_parameters(self):
+        """
+        Limit search to get fresh data and avoid overloading the court server
+        Default: last month
+        """
+        date_filter_end = datetime.today()
+        date_filter_start = date_filter_end - timedelta(30)
+        self.parameters = {
+            "params": {
+                "opn": "",
+                "field_opn_short_title_value": "",
+                "field_opn_issdate_value[min][date]": date_filter_start.strftime(
+                    "%m/%d/%Y"
+                ),
+                "field_opn_issdate_value[max][date]": date_filter_end.strftime(
+                    "%m/%d/%Y"
+                ),
+            }
+        }
+
+    def _download_backwards(self, page_number: int) -> None:
+        self.parameters = {}  # delete date filters used in normal scraper
+        self.url = (
+            f"{self.base_url}?page={page_number}"
+            if page_number > 0
+            else self.base_url
+        )
+        self.html = self._download()
+        self._process_html()
+
+    def build_full_docket_number(
+        self, docket_number: str, lower_court: str
+    ) -> str:
+        """
+        Full docket number has the lower court abbreviation in it
+        For each unique lower court in the opinions available on the source,
+        the linked opinion PDF was opened and the abbrevation extracted
+        """
+        lower_court_abbreviation = self.lower_court_to_abbreviation.get(
+            lower_court, ""
+        )
+
+        return f"BAP No. {lower_court_abbreviation} {docket_number}"
+
+    @staticmethod
+    def get_status_from_opinion_name(opinion_name: str) -> str:
+        if opinion_name.endswith("U"):
+            status = "Unpublished"
+        elif opinion_name.endswith("P"):
+            status = "Published"
+        else:
+            status = "Unknown"
+
+        return status
```

Comment:
Actually  this is close - but occasionally 

`[01-045P1](https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/01-045P1.pdf)` or this field doesnt end in the key and looks like this.

Reply:
I checked the opinions with names ending on P1, only found a couple that were tagged explicitly as "FOR PUBLICATION"
[Example 1](https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/04-029P1.pdf)
[Example 2](https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/97-021P1.pdf)

Also, I found an example labelled which a name ending with [P2](https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/02-079P2.pdf), which is also tagged explicitly "FOR PUBLICATION" 

and an example which ends with ["U1"](https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/02-079U1.pdf) 

Finally, there is another format [03-029P1.01A](https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/03-029P1.pdf) which is also tagged explicitly "FOR PUBLICATION" (there are a bunch around page 17). And the equivalent with an "U" [03-023U1.01A](https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/03-023U1.pdf)

So, should we assume that all opinions ending with the "P" formats are published and all those ending with the "U" formats are unpublished?

Reply:
Related, there is a third distinct format ending with an "O" and a number. From skimming, the BAP seems to be ordering something. Is there a status related?

https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/97-072O1.pdf
https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/97-072O2.pdf
https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/01-027O1.pdf
https://www.bap1.uscourts.gov/sites/bap1/files/bap.pdf.opinions/96-008O1.pdf

Reply:
Yes.  I think U means unpublished. P means published and o means orders which we shouldn't collect.

Reply:
Just tested this for the status if you want to add it
```{python}
    @staticmethod
    def get_status_from_opinion_name(opinion_name: str) -> str:
        if re.search(r"U\d?(\.0\dA)?$", opinion_name):
            status = "Unpublished"
        elif re.search(r"P\d?(\.0\dA)?$", opinion_name):
            # Should match: 02-084P1.01A, 00-094P1, 21-021P
            status = "Published"
        elif re.search(r"O\d$", opinion_name):
            raise SkipOrderError(opinion_name)
        else:
            status = "Unknown"

        return status
```

and it would need to add the following:
```{python}
import re
from juriscraper.AbstractSite import logger


class SkipOrderError(Exception):
    def __init__(self, opinion_name: str):
        logger.debug(f"Skipping {opinion_name} because it is an order")
```

```{python}
    def _process_html(self) -> None:
        for row in self.html.xpath("//tr[td]"):
            opinion_name = row.xpath("td[1]")[0].text_content().strip()

            try:
                status = self.get_status_from_opinion_name(opinion_name)
            except SkipOrderError:
                continue

            lower_court = row.xpath("td[4]/span")[0].text_content().strip()
            "..."
```